testscpp/tests/test_quadratic_model.cpp

```diff
@@ -19,567 +19,58 @@
-            }
-
-            THEN("we can modify the biases with []") {
-                neighborhood[0] += 7;
-                neighborhood[2] -= 3;
-
-                REQUIRE(neighborhood.at(0) == 7.5);
-                REQUIRE(neighborhood.at(2) == -3);
-            }
-
-            THEN("we can create a vector from the neighborhood") {
-                std::vector<std::pair<size_t, float>> pairs(neighborhood.begin(),
-                                                            neighborhood.end());
-
-                REQUIRE(pairs[0].first == 0);
-                REQUIRE(pairs[0].second == .5);
-                REQUIRE(pairs[1].first == 1);
-                REQUIRE(pairs[1].second == 1.5);
-                REQUIRE(pairs[2].first == 3);
-                REQUIRE(pairs[2].second == -3);
-            }
-
-            THEN("we can create a vector from the const neighborhood") {
-                std::vector<std::pair<size_t, float>> pairs(neighborhood.cbegin(),
-                                                            neighborhood.cend());
-
-                REQUIRE(pairs[0].first == 0);
-                REQUIRE(pairs[0].second == .5);
-                REQUIRE(pairs[1].first == 1);
-                REQUIRE(pairs[1].second == 1.5);
-                REQUIRE(pairs[2].first == 3);
-                REQUIRE(pairs[2].second == -3);
-            }
-
-            THEN("we can modify the biases via the iterator") {
-                auto it = neighborhood.begin();
-
-                (*it).second = 18;
-                REQUIRE(neighborhood.at(0) == 18);
-
-                it++;
-                (*it).second = -48;
-                REQUIRE(neighborhood.at(1) == -48);
-
-                ++it;
-                it->second = 104;
-                REQUIRE(neighborhood.at(3) == 104);
-            }
-
-            THEN("we can erase some with an iterator") {
-                neighborhood.erase(neighborhood.begin() + 1, neighborhood.end());
-                REQUIRE(neighborhood.size() == 1);
+// todo: reorganize other tests
+TEST_CASE("QuadraticModel tests") {
+    GIVEN("a linear QM with mixed vartypes") {
+        auto qm = QuadraticModel<double>();
+        qm.add_variable(Vartype::BINARY);
+        qm.add_variable(Vartype::INTEGER, -1, 1);
+        qm.add_variable(Vartype::REAL, -2, 2);
+        qm.add_variable(Vartype::REAL, -3, 3);
+        qm.add_variable(Vartype::SPIN);
+        qm.set_offset(5);
+        qm.set_linear(0, {0, -1, +2, -3, +4});
+
+        WHEN("we use remove_variable()") {
+            qm.remove_variable(2);
+
+            THEN("the variable is removed and the model is reindexed") {
+                REQUIRE(qm.num_variables() == 4);
+                REQUIRE(qm.num_interactions() == 0);
+                CHECK(qm.offset() == 5);
+                CHECK(qm.linear(0) == 0);
+                CHECK(qm.linear(1) == -1);
+                CHECK(qm.linear(2) == -3);  // this was reindexed
```

Comment:
There isn't a way to avoid it, because we're deleting a variable in the middle of the model. The performance of deleting is pretty bad, but luckily it's not used very much.